src/app/downloads/downloads-chart.component.ts

```diff
@@ -22,74 +22,68 @@ export class DownloadsChartComponent implements OnDestroy {
   podcastMetricsStoreSub: Subscription;
   podcastMetrics: PodcastMetricsModel;
   episodeMetricsStoreSub: Subscription;
-  episodeMetrics: EpisodeMetricsModel[];
+  episodeMetrics: EpisodeMetricsModel[] = [];
   episodeChartData: TimeseriesChartModel[];
   podcastChartData: TimeseriesChartModel;
   chartData: TimeseriesChartModel[];
   colors: string[];
 
   constructor(public store: Store<any>) {
     this.filterStoreSub = store.select(selectFilter).subscribe((newFilter: FilterModel) => {
-      if (this.isPodcastChanged(newFilter)) {
-        // reset episode metrics if the filtered podcast changes
-        this.episodeChartData = [];
-      }
-      // this bit picks up changes when episodes are removed from the filter
-      if (this.isEpisodesChanged(newFilter)) {
-        this.episodeMetrics = filterEpisodeMetrics(newFilter, this.episodeMetrics, 'downloads');
-        this.buildEpisodeMetrics(newFilter);
-      }
       this.filter = newFilter;
+      this.applyFilterToExistingData();
     });
 
     this.podcastMetricsStoreSub = store.select(selectPodcastMetrics).subscribe((podcastMetrics: PodcastMetricsModel[]) => {
-      this.podcastMetrics = filterPodcastMetrics(this.filter, podcastMetrics);
-      if (this.podcastMetrics) {
-        this.podcastChartData = this.mapPodcastData(metricsData(this.filter, this.podcastMetrics, 'downloads'));
-        this.updateChartData();
-      }
+      this.updatePodcastChartData(podcastMetrics);
     });
 
     if (!this.episodeMetricsStoreSub) {
       this.episodeMetricsStoreSub = store.select(selectEpisodeMetrics).subscribe((episodeMetrics: EpisodeMetricsModel[]) => {
         this.episodeMetrics = filterEpisodeMetrics(this.filter, episodeMetrics, 'downloads');
-        this.buildEpisodeMetrics(this.filter);
+        this.updateEpisodeChartData();
       });
     }
   }
 
-  buildEpisodeMetrics(filter: FilterModel) {
-    if (this.episodeMetrics) {
-      this.colors = generateShades(this.episodeMetrics.length);
-      this.episodeChartData = this.episodeMetrics
-        .sort((a: EpisodeMetricsModel, b: EpisodeMetricsModel) => {
-          return getTotal(metricsData(filter, b, 'downloads')) - getTotal(metricsData(filter, a, 'downloads'));
-        })
-        .map((metrics: EpisodeMetricsModel, i) => {
-          const episode = filter.episodes.find(ep => ep.id === metrics.id);
-          return this.mapEpisodeData(episode, metricsData(filter, metrics, 'downloads'), this.colors[i]);
-        });
+  applyFilterToExistingData() {
```

Comment:
this is nice!

Reply:
thanks, it is definitely getting better! 😸